src/main/java/com/yelp/nrtsearch/server/luceneserver/NRTPrimaryNode.java

```diff
@@ -122,21 +127,24 @@ static class MergePreCopy {
     final Set<ReplicationServerClient> connections = Collections.synchronizedSet(new HashSet<>());
     final Map<ReplicationServerClient, Iterator<TransferStatus>> clientToTransferStatusIterator;
     final Map<String, FileMetaData> files;
+    private final Deadline deadline;
     private boolean finished;
 
     public MergePreCopy(
         Map<String, FileMetaData> files,
-        Map<ReplicationServerClient, Iterator<TransferStatus>> clientToTransferStatusIterator) {
+        Map<ReplicationServerClient, Iterator<TransferStatus>> clientToTransferStatusIterator,
+        Deadline deadline) {
       this.files = files;
       this.clientToTransferStatusIterator = new ConcurrentHashMap<>(clientToTransferStatusIterator);
       this.connections.addAll(clientToTransferStatusIterator.keySet());
+      this.deadline = deadline;
     }
 
     public synchronized boolean tryAddConnection(
         ReplicationServerClient c, String indexName, long primaryGen, FilesMetadata filesMetadata) {
-      if (!finished) {
+      if (!finished && (deadline == null || !deadline.isExpired())) {
         Iterator<TransferStatus> transferStatusIterator =
-            c.copyFiles(indexName, primaryGen, filesMetadata);
+            c.copyFiles(indexName, primaryGen, filesMetadata, deadline);
```

Comment:
I'm not sure if this is the right logic here. There is still a race condition where the deadline can expire before the request is initiated. What is the behavior if this happens, does it immediately throw an exception or would it happen when the iterator is consumed? If it is the former, we would probably want to catch and ignore the exception here, so it doesn't bubble up.

Reply:
@aprudhomme I tested this particular condition by modifying `ReplicationServerTest.copyFiles`. The error is thrown when the iterator is consumed:
```
DEADLINE_EXCEEDED: ClientCall started after deadline exceeded. Deadline exceeded after -1.004448333s. 
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: ClientCall started after deadline exceeded. Deadline exceeded after -1.004448333s. 
	at app//io.grpc.Status.asRuntimeException(Status.java:535)
	at app//io.grpc.stub.ClientCalls$BlockingResponseStream.hasNext(ClientCalls.java:660)
	at app//com.yelp.nrtsearch.server.grpc.ReplicationServerTest.copyFiles(ReplicationServerTest.java:151)
```
You can see that even though the error says `ClientCall started after deadline exceeded` the error was thrown in `hasNext()`

Reply:
With that behavior, a try-catch shouldn't be needed